### PR TITLE
refactor: prefer patterns for JSON and CLI decoding

### PR DESCRIPTION
**Pull request by bobzhang**
## Summary
- refactor `Usage::from_json` into one optional-field object pattern
- default only absent counters to zero; reject malformed counters, `null`, and non-object values with `JsonDecodeError`
- prefer direct, nested, and multi-field map patterns for fixed JSON keys across DeepSeek, session persistence, auto-check, multi-edit, plan, read, and visualization packages
- decode fixed `@argparse.Matches` values, flags, and sources with record/map patterns in `openseek`, TUI, visualizer, and shared agent CLI configuration
- retain `.get(name)` and shared accessors only for genuinely dynamic keys
- add focused tests for malformed Usage fields, non-object input, and present-null response usage

## CLI pattern scope
- TUI configuration matches required defaults and optional engine/session fields together
- visualizer startup converts one `Matches` value into a typed `VizConfig`
- session compaction matches `file`, `from`, and `to` in one branch
- model/thinking decoding is shared across run, fleet, and serve modes
- visualizer search behavior matches `values` and `sources` together, preserving explicit flag/env precedence

`with` is used for scalar defaults. Constructed collection defaults and read-only external enum constructors cannot be used as `with` defaults, so those cases remain ordered pattern arms.

## Scope
This completes the root production-code sweep for fixed literal JSON and CLI keys outside the desktop, eval, and test-heavy trees. The refactor changes no public interfaces.

## Pattern reachability
A `moon run --deny-warn -e` probe confirmed that a direct `Map` pattern followed by a redundant wildcard reports `unreachable_code`. The corresponding `Json` pattern does not warn because its wildcard still handles non-object variants, so no core issue was filed.

## GitHub review feedback
The streaming decoder intentionally discards wire-level `"usage": null` placeholders before assembling a `ChatResponse`. Direct `ChatResponse` decoding rejects a present null value, consistent with the contract that only an absent field defaults; a regression test documents this distinction and the review thread is resolved.

## Validation
- `moon info` (no public interface changes)
- `moon fmt --check`
- `moon check --deny-warn`
- affected CLI packages: 168 tests
- `moon test` (1,048 native tests)
- `moon test --target js` (228 tests)
- `moon cram test tests/cram` (23 tests)
- fresh ephemeral, read-only Codex CLI review of the complete final PR: no actionable regressions

### CODE REVIEW

File: deepseek/json_decode.mbt [hunk 1]
```diff
@@ -94,11 +102,9 @@ pub impl FromJson for ChatResponse with fn from_json(
   let (message_fields, usage_json) : (Map[String, Json], Json?) = match json {
     {
       "choices": [{ "message": Object(message_fields), .. }, ..],
-      "usage": usage_json,
+      "usage"? : usage_json,
       ..
-    } => (message_fields, Some(usage_json))
-    { "choices": [{ "message": Object(message_fields), .. }, ..], .. } =>
-      (message_fields, None)
+    } => (message_fields, usage_json)
```
[the remark this answers is not in the document]
Comment (bobzhang) [reply deepseek/json_decode.mbt:107]: Thanks. This is intentionally strict per the decoder contract: only a missing `usage` field defaults; any present non-object value, including `null`, is a format error. The streaming path is distinct: `apply_stream_json` explicitly ignores `{"usage": null}` placeholders before `StreamAccumulator::response` constructs the final envelope, so those wire-level chunks never reach `ChatResponse::from_json`. I added `chat response decode rejects present null usage` in 570243a5 to document the direct-decoder behavior. The full suite and a fresh Codex CLI review are clean.
